iothub_client/samples/pnp/pnp_temperature_controller/pnp_temperature_controller.c

```diff
@@ -0,0 +1,381 @@
+        size_t jsonToSendStrLen = strlen(jsonToSendStr);
+
+        if ((iothubClientResult = IoTHubDeviceClient_SendReportedState(deviceClient, (const unsigned char*)jsonToSendStr, jsonToSendStrLen, NULL, NULL)) != IOTHUB_CLIENT_OK)
+        {
+            LogError("Unable to send reported state, error=%d", iothubClientResult);
+        }
+        else
+        {
+            LogInfo("Sending serialNumber property to IoTHub");
+        }
+    }
+
+    STRING_delete(jsonToSend);
+}
+
+
+//
+// CreateDeviceClientAndAllocateComponents allocates IOTHUB_DEVICE_CLIENT_HANDLE the application will use along with thermostat components
+// 
+static IOTHUB_DEVICE_CLIENT_HANDLE CreateDeviceClientAndAllocateComponents(void)
+{
+    IOTHUB_DEVICE_CLIENT_HANDLE deviceClient = NULL;
+    const char* connectionString;
+    bool result;
+
+    if ((connectionString = getenv(g_connectionStringEnvironmentVariable)) == NULL)
+    {
+        LogError("Cannot read environment variable=%s", g_connectionStringEnvironmentVariable);
+        result = false;
+    }
+    else if ((deviceClient = PnPHelper_CreateDeviceClientHandle(connectionString, g_thermostatModelId, g_hubClientTraceEnabled, PnP_TempControlComponent_DeviceMethodCallback, PnP_TempControlComponent_DeviceTwinCallback)) == NULL)
+    {
+        LogError("Failure creating IotHub device client");
+        result = false;
+    }
+    else if ((g_thermostatHandle1 = PnP_ThermostatComponent_CreateHandle(g_thermostatComponent1Name)) == NULL)
+    {
+        LogError("Unable to create component handle for %s", g_thermostatComponent1Name);
+        result = false;
+    }
+    else if ((g_thermostatHandle2 = PnP_ThermostatComponent_CreateHandle(g_thermostatComponent2Name)) == NULL)
+    {
+        LogError("Unable to create component handle for %s", g_thermostatComponent2Name);
+        result = false;
+    }
+    else
+    {
+        result = true;
+    }
+
+    if (result == false)
+    {
+        PnP_ThermostatComponent_Destroy(g_thermostatHandle2);
+        PnP_ThermostatComponent_Destroy(g_thermostatHandle1);
+        if (deviceClient != NULL)
+        {
+            IoTHubDeviceClient_Destroy(deviceClient);
+            IoTHub_Deinit();
+            deviceClient = NULL;
+        }
+    }
+
+    return deviceClient;
+}
+
+int main(void)
+{
+    IOTHUB_DEVICE_CLIENT_HANDLE deviceClient = NULL;
+
+    if ((deviceClient = CreateDeviceClientAndAllocateComponents()) == NULL)
+    {
+        LogError("Failure creating IotHub device client");
+    }
+    else
+    {
+        LogInfo("Successfully created device client.  Hit Control-C to exit program\n");
+
+        // During startup, send the non-"writeable" properties.
+        PnP_TempControlComponent_ReportSerialNumber_Property(deviceClient);
+        PnP_DeviceInfoComponent_Report_All_Properties(deviceClient, g_deviceInfoComponentName);
+        PnP_TempControlComponent_Report_MaxTempSinceLastReboot_Property(g_thermostatHandle1, deviceClient);
```

Comment:
Ugh it's definitely not consistent.  Cleaning up.

Reply:
The handle should actually come 1st.  This follows convention of rest of SDK where handle comes 1st.  The issue is that the deviceInfo & temperatureController don't need handles because they don't actually allocate memory and I'm trying to keep this as small as possible and only malloc where needed.  (I'd considered having a dummy handle for deviceInfo/tempController - so like `#define DEVICE_INFO_HANDLE (0x12345)` to keep it consistent, so it'd just be some magic # that got ignored, but I worry that's more likely to confuse since the handle business in tempController already adds some extra logic here.)  

The other potential option if the handle business is too complex is to just say forget abstraction and have the `struct PNP_THERMOSTAT_COMPONENT` defined in `pnp_thermostat_component.h` so it's just a struct.  @danewalton you've stared at this recently, and specifically your comment RE `This function says TempControlComponent which leads me to believe it is a subcomponent. If that's true, should the NULL param be the component's name?` points to places to clarify.  Do you have thoughts about just saying no handles, all structs for components (that the main() would still NOT directly access, other than knowing how much space to allocate to store on the .bss).  So we'd have 

```c
// pnp_thermostat.h
struct PNP_THERMOSTAT_COMPONENT {...}
PnP_ThermostatComponent_Initialize(PNP_THERMOSTAT_COMPONENT *p,const char* componentName); // replaces PnP_ThermostatComponent_CreateHandle.  

// pnp_temperature_controller.c
PNP_THERMOSTAT_COMPONENT g_thermostat1; // static allocation at top of file

```

Reply:
OK - the issue here is that I want to keep sample simple, but I also want to model abstracting and specifically composition of multiple components where we don't have data structure details poking out in a header.

My inclination is to leave code as-is, except for some of the consistency cleanup that Rido had initially called out that is indeed legit.